NDTensors/src/lib/Sectors/src/category_product.jl

```diff
@@ -34,78 +45,206 @@ function Base.show(io::IO, s::CategoryProduct)
+)
+  return sector()
+end
+
+# EmptyCategory acts as trivial on any AbstractCategory, not just CategoryProduct
+function fusion_rule(::SymmetryStyle, ::CategoryProduct{Tuple{}}, c::AbstractCategory)
+  return to_graded_axis(c)
+end
+function fusion_rule(::SymmetryStyle, ::CategoryProduct{Tuple{}}, c::CategoryProduct)
+  return to_graded_axis(c)
+end
+function fusion_rule(::SymmetryStyle, c::AbstractCategory, ::CategoryProduct{Tuple{}})
+  return to_graded_axis(c)
+end
+function fusion_rule(::SymmetryStyle, c::CategoryProduct, ::CategoryProduct{Tuple{}})
+  return to_graded_axis(c)
+end
+
+# abelian case: return Category
+fusion_rule(::AbelianGroup, ::CategoryProduct{Tuple{}}, c::AbstractCategory) = c
+fusion_rule(::AbelianGroup, ::CategoryProduct{Tuple{}}, c::CategoryProduct) = c
+fusion_rule(::AbelianGroup, c::AbstractCategory, ::CategoryProduct{Tuple{}}) = c
+fusion_rule(::AbelianGroup, c::CategoryProduct, ::CategoryProduct{Tuple{}}) = c
+
+# ===============================  Ordered implementation  =================================
+CategoryProduct(t::Tuple) = _CategoryProduct(t)
+CategoryProduct(cats::AbstractCategory...) = CategoryProduct((cats...,))
+
+categories_isequal(o1::Tuple, o2::Tuple) = (o1 == o2)
+
+categories_isless(::Tuple, ::Tuple) = throw(ArgumentError("Not implemented"))
+categories_isless(t1::T, t2::T) where {T<:Tuple} = t1 < t2
 
+categories_product(l1::Tuple, l2::Tuple) = (l1..., l2...)
+
+categories_trivial(type::Type{<:Tuple}) = trivial.(fieldtypes(type))
+
+function find_common(t1::Tuple, t2::Tuple)
+  n = min(length(t1), length(t2))
+  return t1[begin:n], t2[begin:n]
+end
+
+function find_diff(t1::Tuple, t2::Tuple)
+  n1 = length(t1)
+  n2 = length(t2)
+  return n1 < n2 ? t2[(n1 + 1):end] : t1[(n2 + 1):end]
+end
+
+# ===========================  Dictionary-like implementation  =============================
 function CategoryProduct(nt::NamedTuple)
   categories = sort_keys(nt)
   return _CategoryProduct(categories)
 end
 
 CategoryProduct(; kws...) = CategoryProduct((; kws...))
 
+# avoid having 2 different kinds of EmptyCategory: cast empty NamedTuple to Tuple{}
+CategoryProduct(::NamedTuple{()}) = CategoryProduct(())
+
 function CategoryProduct(pairs::Pair...)
   keys = ntuple(n -> Symbol(pairs[n][1]), length(pairs))
   vals = ntuple(n -> pairs[n][2], length(pairs))
   return CategoryProduct(NamedTuple{keys}(vals))
 end
 
-function categories_fusion_rule(A::NamedTuple, B::NamedTuple)
-  qs = [A]
-  for (la, lb) in zip(pairs(intersect_keys(A, B)), pairs(intersect_keys(B, A)))
-    @assert la[1] == lb[1]
-    fused_vals = ⊗(la[2], lb[2])
-    qs = [union_keys((; la[1] => v), q) for v in fused_vals for q in qs]
-  end
-  # Include sectors of B not in A
-  qs = [union_keys(q, B) for q in qs]
-  return qs
+# sector() acts as empty NamedTuple
+function categories_isequal(nt::NamedTuple, ::Tuple{})
+  return categories_isequal(nt, categories_trivial(typeof(nt)))
```

Comment:
```suggestion
  return categories_isequal(nt, categories_trivial(nt))
```
May as well define `categories_trivial` for instances along with types and simplify the calls to `categories_trivial`.

Reply:
Alternatively, can you just define it as:
```julia
function categories_isequal(nt::NamedTuple, ::Tuple{})
  return categories_isequal(nt, (;))
end
```
and then the trivial categories can be filled in by the generic function:
```julia
function categories_isequal(nt1::NamedTuple, nt2::NamedTuple)
  return ==(sym_categories_insert_unspecified(nt1, nt2)...)
end
```
defined below?